nixos/tests/nix/upgrade.nix

```diff
@@ -0,0 +1,111 @@
+{ pkgs, ... }:
+let
+  lib = pkgs.lib;
+
+  fallback-paths-external = pkgs.writeTextDir "fallback-paths.nix" ''
+    {
+      ${pkgs.system} = "${pkgs.nixVersions.latest}";
+    }'';
+
+  inputDrv = import ../.. {
+    configuration =
+      { pkgs, modulesPath, ... }:
+      {
+        imports = [ nixos-module ];
+        nix.package = pkgs.nixVersions.latest;
+        boot.isContainer = true;
+      };
+    system = pkgs.system;
+  };
+
+  nixos-module = builtins.toFile "nixos-module.nix" ''
+    { lib, pkgs, modulesPath, ... }:
+    {
+      imports = [
+        (modulesPath + "/profiles/minimal.nix")
+        (modulesPath + "/testing/test-instrumentation.nix")
+      ];
+
+      hardware.enableAllFirmware = lib.mkForce false;
+
+      nix.settings.substituters = lib.mkForce [];
+      nix.settings.hashed-mirrors = null;
+      nix.settings.connect-timeout = 1;
+      nix.extraOptions = "experimental-features = nix-command";
+
+      environment.localBinInPath = true;
+      users.users.alice = {
+        isNormalUser = true;
+        packages = [ pkgs.nixVersions.latest ];
+      };
+      documentation.enable = false;
+    }
+  '';
+in
+
+pkgs.testers.nixosTest rec {
+  name = "nix-upgrade-${pkgs.nixVersions.stable.version}-${pkgs.nixVersions.latest.version}";
+  meta.maintainers = with lib.maintainers; [ tomberek ];
+
+  nodes.machine =
+    { lib, modulesPath, ... }:
+    {
+      imports = [ nixos-module ];
+
+      virtualisation.memorySize = 9192;
+      virtualisation.cores = 4;
```

Comment:
Do you need that much cores or would you leave it to the defaults?

Reply:
Not really, it helped when I was iterating to do the full eval + build inside the QEMU. Removing...